File: banco_problemas_aplicados/nivel_intermedio/soluciones/c/problema26.c
/*
 * Problema 26 - Desplazamiento Circular de Muestras
 * void rotaDerecha(double Arr[], int tam, int k)
 * Contexto: simulacion de retardos. Rotacion ciclica hacia la derecha por K posiciones:
 * el elemento de la posicion i pasa a la posicion (i + k) % tam. Lo que sale por la derecha
 * reingresa por la izquierda. (Ver teoria/03_transformaciones_de_vector.md)
 */
#include <stdio.h>

void rotaDerecha(double arr[], int tam, int k) {
    k = k % tam;                        /* si k >= tam, la rotacion se repite */
    double aux[tam];
    for (int i = 0; i < tam; i++) {
        aux[(i + k) % tam] = arr[i];    /* indice circular con modulo */
    }
    for (int i = 0; i < tam; i++) {
        arr[i] = aux[i];                /* copiar el resultado de vuelta */
    }
}

int main(void) {
    int tam, k;
    printf("Ingrese el numero de muestras: ");
    scanf("%d", &tam);

    double arr[tam];
    for (int i = 0; i < tam; i++) {
        printf("Muestra %d: ", i);
        scanf("%lf", &arr[i]);
    }

    printf("Ingrese K (posiciones a rotar a la derecha): ");
    scanf("%d", &k);

    rotaDerecha(arr, tam, k);

    printf("Buffer desplazado: ");
    for (int i = 0; i < tam; i++) {
        printf("%.2f ", arr[i]);
    }
    printf("\n");
    return 0;
}

File: banco_problemas_aplicados/nivel_intermedio/soluciones/java/Problema26.java
/*
 * Problema 26 - Desplazamiento Circular de Muestras
 * void rotaDerecha(double[] arr, int tam, int k)  -> rota a la derecha K posiciones (ciclico).
 */
import java.util.Scanner;
import java.util.Locale;

public class Problema26 {

    static void rotaDerecha(double[] arr, int tam, int k) {
        k = k % tam;
        double[] aux = new double[tam];
        for (int i = 0; i < tam; i++) {
            aux[(i + k) % tam] = arr[i];
        }
        for (int i = 0; i < tam; i++) {
            arr[i] = aux[i];
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Ingrese el numero de muestras: ");
        int tam = sc.nextInt();

        double[] arr = new double[tam];
        for (int i = 0; i < tam; i++) {
            System.out.print("Muestra " + i + ": ");
            arr[i] = sc.nextDouble();
        }

        System.out.print("Ingrese K (posiciones a rotar a la derecha): ");
        int k = sc.nextInt();

        rotaDerecha(arr, tam, k);

        System.out.print("Buffer desplazado: ");
        for (int i = 0; i < tam; i++) {
            System.out.printf(Locale.US, "%.2f ", arr[i]);
        }
        System.out.println();
        sc.close();
    }
}
